[Enhancement] Time Splits

Caladius opened the pull request:
Not really a port over from SoH, this was built from the ground up for 2Ship with some elements borrowed from the original. Did my damnest to reduce the footprint outside of my added files.

Would appreciate testing to see what Items don't split and any bugs encountered. 

- New Enhancement Sidebar Menu "Time Splits".
- Adds Inventory, Masks, Songs, and Quest Items.
- Preview List allows for Click-to-Remove and Drag & Drop Reordering.
- Clicking on an Icon in the Overlay List will Skip it.
- Adds a new GI Hook for when Bottle Contents update in the inventory (Swing to collect).
- Adds a new GI Hook for when Dungeon Bosses are defeated (Boss Kills are handled so inconsistently in this game).

To-do:
- [x] Acquiring Magic Meters (both Single and Double Meters)
- [x] Double Defense
- [x] Saving/Loading Split Lists from File
- [x] Dungeon Boss Kills (Odolwa, Goht, etc.)

Future To-do (Potential V2, you know I love a good V2...):
- [x] Entrances (Entering a zone for the first time)

These were in the initial list for a V2 but will probably be added to a separate subsystem and not directly integrated with Splits:
- Rando Support (Split items out of order, akin to SoH Gameplay Stats)
- Custom Icons for Boss Kill splits.
- Dungeon Items (Keys, Stray Fairies)

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4278305467.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4278412124.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/4278927928.zip)
<!--- section:artifacts:end -->

Caladius committed:
Adds support for Timesplits

Tristnal commented:
Client crash upon enabling Time Splits in the popped out window, or by clicking the button to pop out the window with the "Enable Time Splits" option checked.
[2 Ship 2 Harkinian.log](https://github.com/user-attachments/files/22192176/2.Ship.2.Harkinian.log)

Caladius committed:
clanged

Caladius committed:
Fix Popout Crash

Caladius commented:
> Client crash upon enabling Time Splits in the popped out window, or by clicking the button to pop out the window with the "Enable Time Splits" option checked. [2 Ship 2 Harkinian.log](https://github.com/user-attachments/files/22192176/2.Ship.2.Harkinian.log)

Pushed up the fix for that, thanks. When its done building you can snag a new build.

Tristnal commented:
Bow split doesn't seem to work when obtaining the bow/first quiver in Woodfall.

Caladius commented:
I was curious about that. Didn't know if it gave an actual bow or the quiver.

Thanks, I'll fix it tomorrow

sigmin102 commented:
I see that defeat bosses might not be added until later, however I couldn't find a good option to put for the final split of a run. If it'd be possible to have a Defeat Majora split similar to Defeat Ganon from OoT, it'd be nice to have a final split. The timer I believe would still stop when Majora is defeated (I'll test that soon).

Additionally, if there could be a way to resize the splits on the overlay, similar to Ship, that'd be really great for keeping the screen cleaner. I have a lot of splits for Glitchless Any% and could simply reduce the number of splits to work around this.

sigmin102 commented:
I tested this out with a Glitchless Any% run (using dev tools warps, but the same route I'd take) and here's what I found:

1. Red Potion Bottle from Kotake and Big Poe in bottle did not trigger their bottle split
2. Gold Dust and Full Milk (from Aliens) did trigger their bottle split
3. The main timer ended on my last split, so not on Majora. My last split was Twinmold remains.
4. Boss remains splits were upon entering Blue Warp instead of when Link holds the Boss Remains over his head (this is fine imo)
5. Adding/Removing splits during a run attempt worked (that's awesome)
6. Splits were remembered after Owl Save back to Title Screen and re-entering File (awesome)
7. All other splits besides the Red Potion and Big Poe worked flawlessly. Obvious workaround is just not enabling Red Potion/Big Poe. Other bottle contents may not work as well, didn't exhaustively test.

I think this is awesome so far, thank you for your work on this! I eagerly await other planned functionality.

balloondude2 commented on mm/2s2h/Enhancements/Trackers/TimeSplits/TimeSplitsActions.cpp:
```diff
@@ -0,0 +1,199 @@
+    // });
+
+    COND_HOOK(OnItemGive, CVAR, [](u8 item) {
+        if (item == ITEM_HEART_PIECE_2) {
+            item = ITEM_HEART_PIECE;
+        }
```
Internally, the Hag's red potion botte is given with `ITEM_LONGSHOT`, so you might want to add another conversion here for that, or implement some other way to associate `ITEM_LONGSHOT` with the Red Bottle split.

Caladius replied:
Thanks for checking! I fixed bow already, and I have a hunch on big poe

Caladius replied:
This is now fixed, thanks Bdude

Caladius committed:
wip file saving; fixes Bow and Red Potion splits

Caladius committed:
Add Active Row Highlighting

Caladius committed:
Fix D-Pad Bottle Collection

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.cpp:
```diff
@@ -0,0 +1,229 @@
+ImGuiTableFlags tableColumnFlags = ImGuiTableColumnFlags_None;
+ImVec4 splitOpacity = { 0, 0, 0, 0.5f };
+
+std::string formatTimesplitTime(uint32_t value) {
+    uint32_t sec = value / 10;
+    uint32_t hh = sec / 3600;
+    uint32_t mm = (sec - hh * 3600) / 60;
+    uint32_t ss = sec - hh * 3600 - mm * 60;
+    uint32_t ds = value % 10;
+    return fmt::format("{}:{:0>2}:{:0>2}.{}", hh, mm, ss, ds);
+}
```
DisplayOverlay.cpp defines the same sort of time string in `formatTimeDisplay`. Might be better to commonly define this somewhere.

Caladius replied:
would it make more sense to put it in ShipUtils then?

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.cpp:
```diff
@@ -0,0 +1,229 @@
+            // Item Image Column
+            ImGui::TableNextColumn();
+
+            if (CVarGetInteger("gSettings.TimeSplits.Highlight", 0) && splits.splitStatus == SPLIT_ACTIVE) {
```
I see a number of CVar names in this PR that are only referenced once. Are they WIP?

Caladius replied:
pretty sure all of them are only referenced once, they are all toggled checkboxes that need their state saved/restored on relaunch of the game

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/TimesplitsSettings.cpp:
```diff
@@ -0,0 +1,381 @@
+#include "assets/objects/gameplay_keep/gameplay_keep.h"
+#include "GameInteractor/GameInteractor.h"
+
+std::map<uint32_t, ImVec4> songColorMap = {
+    { ITEM_SONG_SONATA, ImVec4(0.588f, 1.0f, 0.392f, 1.0f) }, { ITEM_SONG_LULLABY, ImVec4(1.0f, 0.313f, 0.156f, 1.0f) },
+    { ITEM_SONG_NOVA, ImVec4(0.392f, 0.588f, 1.0f, 1.0f) },   { ITEM_SONG_ELEGY, ImVec4(1.0f, 0.627f, 0.0f, 1.0f) },
+    { ITEM_SONG_OATH, ImVec4(1.0f, 0.392f, 1.0f, 1.0f) },
+};
```
At some point I'd like there to be a common reference for these song colors. The Item Tracker already defines these values for example, which your starting items rework PR moved that to ShipUtils. I'm not saying that the onus for this change is on this PR, but I want to voice this concern so that it's in the mindshare.

Caladius replied:
the starting items rework was definitely the goal to move it there for this reason, wasnt sure if i should double that here or leave it for whichever PR gets merged second to clean up.

Caladius replied:
I ended up moving this to ShipUtils under an itemColorMap, Starting Items will need to conflict resolve as this map will be able to house all non Rando items.

Proxy wasn't a fan of mixing RI's and vanilla Item ID's so we will just use the "RI to ItemId" function or a "Quest to ItemId" function when needed to pull from the same list. I also added a Ship_GetItemColorTint function under it that game be used to retrieve an ImVec4 color from it.

Caladius committed:
Cleanup and Add Boss Splits.

Caladius committed:
Add names to Boss Icons

Caladius committed:
clang v2

Caladius committed:
Linker fix

Caladius committed:
Move time formatting to ShipUtils (this also cleaned up DisplayOverlay).

Caladius committed:
clang v3

Caladius committed:
Adds Saving/Loading splits from file.

Caladius committed:
Update inventory range to exclude heart piece.

Caladius committed:
Update column name to not be abbreviated

Caladius committed:
Adds Magic Meters and Double Defense

Caladius committed:
clang cuz ya boi alwayz fogetz

Caladius committed:
Fix Fairy Splits with CS Skip Enabled

Caladius committed:
Move songColorMap to ShipUtils along with the function to get the color. Fixed Song Icon sizing.

Caladius committed:
Adds default Window size and allows resizing.

Caladius committed:
Allow scrolling of splits list

Caladius committed:
Adds Entrance Splits and cleans up the UI.

Adds new splitType to the object so I added a small solution to handle older files that may not have it for testing.

Caladius committed:
UI Update and Entrance Filters.

Caladius committed:
Adds split list to the options window to fill out the space.

Caladius committed:
Fix ID mismatching.

Caladius committed:
Add Split Comparing Functionality

Caladius committed:
clang

Caladius committed:
Compare Split text display corrections

Caladius committed:
Adds auto scrolling to follow Active Splt if list is bigger than the window. Also locks the Header Row when visible to always be visible at the top instead of scrolling with the list.

Caladius committed:
Adds Tooltips for options

Caladius committed:
Fix time coloring for splits with no previous data.

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.cpp:
```diff
@@ -0,0 +1,303 @@
+ImVec4 splitOpacity = { 0, 0, 0, 0.5f };
+
+SplitTextObject GetCurrentTimeTextDisplay(TimesplitObject split) {
+    uint32_t totalTime = ((GetUnixTimestamp() - gSaveContext.save.shipSaveInfo.fileCreatedAt) / 100);
```
With this exact line being defined three times, instead define it once in a `uint32_t getTotalTime()` function and just call that.

Eblo replied:
It's actually repeated more times than that.

Caladius committed:
Fix crash and ID mismatch

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.cpp:
```diff
@@ -0,0 +1,303 @@
+
+#include "Timesplits.h"
+#include <spdlog/fmt/fmt.h>
```
Is this still needed?

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.cpp:
```diff
@@ -0,0 +1,303 @@
+                ImGui::TableNextColumn();
+                TableCellCenteredText(
+                    GetCurrentTimeTextDisplay(splitList[i]).colorDisplay,
+                    !gPlayState ? "--:--:--.-"
```
Similarly, with this string being defined three times, use something like  `#define BLANK_SPLIT  "--:--:--.-"` somewhere and just reference that.

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/Timesplits.h:
```diff
@@ -0,0 +1,86 @@
+    uint32_t endIndex;
+} IndexRangeObject;
+
+extern std::vector<TimesplitObject> splitList;
+extern std::vector<TimesplitObject> comparisonList;
+extern std::vector<TimesplitObject> splitObjectList;
+extern std::vector<TimesplitObject> sceneObjectList;
+extern std::vector<std::string> savedLists;
+extern std::map<uint32_t, std::vector<uint32_t>> itemSubMenuList;
+extern std::map<uint32_t, ImVec4> songColorMap;
+extern uint32_t comparedIndex;
+extern bool shouldPopUpOpen;
+extern TimesplitObject GetSplitObjectById(uint32_t itemId);
+extern ImVec2 GetItemImageSizeById(uint32_t itemId);
+extern void TableCellCenteredText(ImVec4 color, const char* text);
+extern const char* GetItemImageById(uint32_t itemId);
+extern void SplitsPushImageButtonStyle();
+extern void SplitsPopImageButtonStyle();
+extern void HandlePopUpContext(uint32_t popupId);
+extern void HandleDragAndDrop(size_t i);
+extern void UpdateSplitBests();
+extern void UpdateSplitSettings(uint32_t settingName);
+extern void SkipSplitEntry(uint32_t index);
+extern void AddSplitEntryBySceneId(uint32_t sceneId);
+extern void AddSplitEntryById(uint32_t itemId);
+extern void RemoveSplitEntry(uint32_t splitId, uint32_t index);
+extern void SplitSaveFileAction(uint32_t action, std::string listName);
+extern void DrawSplitsList(bool isMain);
+extern void SplitLoadComparisonList();
```
Do these need `extern` if they're in a header?

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/TimesplitsSettings.cpp:
```diff
@@ -0,0 +1,742 @@
+
+#include "TimesplitsSettings.h"
+#include "Timesplits.h"
+#include <spdlog/fmt/fmt.h>
```
Is this still needed?

Eblo commented on mm/2s2h/Enhancements/Trackers/TimeSplits/TimesplitsSettings.cpp:
```diff
@@ -0,0 +1,742 @@
+
+std::vector<TimesplitObject> sceneObjectList = {
+    // Misc. Areas
+    { SCENE_LOST_WOODS, "Lost Woods (Intro)" },
```
Instead of manually defining these scenes, I think you could do some clever trickery with including `scene_table.h` and messing with the defines there. Something like (haven't tested):

```c++
#define DEFINE_SCENE(_name, enumValue, _textId, _drawConfig, _restrictionFlags, _persistentCycleFlags, \
                     _entranceSceneId, _betterMapSelectIndex, humanName)                               \
    { enumValue, humanName },
#define DEFINE_SCENE_UNSET(_enumValue)

std::vector<TimesplitObject> sceneObjectList = {
#include "tables/scene_table.h"
};

#undef DEFINE_SCENE
#undef DEFINE_SCENE_UNSET
```

Sadly this won't help the others like items, but I think the reduced duplication is worth a shot.

Caladius replied:
My only concern here is that I am defining these in a specific Order so it makes sense in the UI. the Scene Table has them in a arbitrary order that makes finding what you want in the UI a horrible experience.

If you have a good way to retain the order I have created I am open to changing it, otherwise I would prefer to not negatively impact the experience.

Eblo replied:
> If you have a good way to retain the order I have created I am open to changing it, otherwise I would prefer to not negatively impact the experience.

Better Map Select, whose display order might fit what is wanted here, addresses this by manually arranging based on the `betterMapSelectIndex`. However, we'd be repeating something either way. I don't think researching an answer to this problem is in the scope of Time Splits, so I'm fine leaving it as you have it for now.

Eblo commented on mm/2s2h/GameInteractor/GameInteractor.cpp:
```diff
@@ -179,6 +179,13 @@ void GameInteractor_ExecuteOnPlayerPostLimbDraw(Player* player, s32 limbIndex) {
 }
 
+void GameInteractor_ExecuteOnBossDefeated(int16_t actorId) {
+    GameInteractor::Instance->ExecuteHooks<GameInteractor::OnBossDefeated>(actorId);
```
For consistency, let's toss in a debug log
```suggestion
    SPDLOG_DEBUG("GameInteractor_ExecuteOnBossDefeated: actorId: {}", actorId);
    GameInteractor::Instance->ExecuteHooks<GameInteractor::OnBossDefeated>(actorId);
```

Eblo commented on mm/2s2h/GameInteractor/GameInteractor.cpp:
```diff
@@ -179,6 +179,13 @@ void GameInteractor_ExecuteOnPlayerPostLimbDraw(Player* player, s32 limbIndex) {
     GameInteractor::Instance->ExecuteHooksForFilter<GameInteractor::OnPlayerPostLimbDraw>(player, limbIndex);
 }
 
+void GameInteractor_ExecuteOnBossDefeated(int16_t actorId) {
```
```suggestion
void GameInteractor_ExecuteOnBossDefeated(s16 actorId) {
```

Eblo commented on mm/2s2h/GameInteractor/GameInteractor.h:
```diff
@@ -730,6 +730,7 @@ void GameInteractor_ExecuteOnActorDraw(Actor* actor);
 void GameInteractor_ExecuteOnActorKill(Actor* actor);
 void GameInteractor_ExecuteOnActorDestroy(Actor* actor);
 void GameInteractor_ExecuteOnPlayerPostLimbDraw(Player* player, s32 limbIndex);
+void GameInteractor_ExecuteOnBossDefeated(int16_t actorId);
```
```suggestion
void GameInteractor_ExecuteOnBossDefeated(s16 actorId);
```

Eblo commented on mm/2s2h/GameInteractor/GameInteractor_HookTable.h:
```diff
@@ -31,6 +31,7 @@ DEFINE_HOOK(OnActorDraw, (Actor * actor))
 DEFINE_HOOK(OnActorKill, (Actor * actor))
 DEFINE_HOOK(OnActorDestroy, (Actor * actor))
 DEFINE_HOOK(OnPlayerPostLimbDraw, (Player * player, s32 limbIndex))
+DEFINE_HOOK(OnBossDefeated, (int16_t actorId))
```
```suggestion
DEFINE_HOOK(OnBossDefeated, (s16 actorId))
```

Caladius committed:
Feedback Updates

Caladius committed:
Extern is needed since my definition is in the cpp file it seems.

Eblo commented on mm/2s2h/BenGui/BenGui.cpp:
```diff
@@ -140,6 +144,13 @@ void SetupGuiElements() {
+    gui->AddGuiWindow(mTimesplitsWindow);
+
+    mTimesplitsSettingsWindow = std::make_shared<TimesplitsSettingsWindow>(
+        "gWindows.Timesplits.Settings", "Time Splits Settings Window", ImVec2(567, 97));
```
Is there any significance to these specific dimensions?

Caladius replied:
Just what the values were when I sized the window after first use. Seemed like a good default.

Eblo reviewed (approved):
Checks out to me code-wise, and it seems functional from what I know about how splits should work. Left a non-blocking question, but otherwise as long as testers are happy with this, then so am I.

Caladius committed:
Update to latest Dev
